### ISSUE TITLE
User Request: Atomics in dash::UnorderedMap

### ISSUE THREAD

**Issue by fuchsto** [dash-project/dash#249]
User request I received by mail:

> I have come across Dash and found it very interesting. This is something that could allow me, for once, to avoid having to redo everything from scratch with MPI.
> 
> I am trying to use an UnorderedMap for my purposes but I am stumbling into difficulties because one works on a local copy that has to be synchronised later. I am not in the usual situation of a halo, for example, where each node works on specific cells that will be updated at a specific time. In my case, any node could be able to create or increment any element of the map. Even worse, several nodes could try to create an element with the same key at the same time. In that case, the logic would be to simply add their values.
> 
> This is for a genomics application where every node is processing part of a genome and we count specific strips of DNA. Each of these strip will correspond to the key of the map and we simply have to count the number of times they appear. There is a huge number of possibilities (up to 10^12 or even more) of those strips but not all of them will be found. Any node could find any strips. So two nodes could find the same strip and try to insert that element in the map, if it was not already there, or to increment it and that operation should be atomic. 
> 
> It is for that reason that the local copy is causing me difficulties. Of course, I do not know well your library. I have just been experimenting with it so I might have missed an obvious way to do it. I would be grateful if you could give me a suggestion or simply tell me that this is a bad idea. I am not asking for technical help but a mere suggestion.
> 
> Thank you!

**Comment by fuchsto** [comment]
My reply:

First, thank you for evaluating DASH for your application.
We highly appreciate feedback from users and will gladly provide technical support.

The "vanilla" configuration of dash::UnorderedMap is designed to provide a distributed key/value store.
If I understand your use case correctly, you use mapped values for shared counters.
So, for a simplified example:

```c++
dash::Array<GATCValue> dna;
dash::UnorderedMap<GATCSequence, int> map;
// ...
// find match in local data segment:
auto matched_strip_begin = std::find(dna.local.begin(),
                                     dna.local.end(),
                                     gatc_match_fun);
// ...
auto matched_dna_strip = GATCSequence(*matched_strip_begin, ...);
// increment global hit count of matched strip:
map[matched_dna_strip]++; // <-- not atomic
```

(Let me know if I got your use case wrong.)

In DASH, we try to mimic thread-safety and consistency semantics of the STL as closely as possible.
The consistency violation you observed is caused by non-atomic changes to the mapped values. Much like in STL containers, changes to *values* of container elements do not affect container semantics
(opposed to adding/removing container elements).
In conventional (non-DASH) C++, you would use:

```c++
std::unordered_map<GATCSequence, std::atomic<int>>
```

So actually the problem is the container value type. In your case, you would need something like:

```c++
dash::UnorderedMap< GATCSequence, dash::Atomic<int> > map;
```

However, I'm afraid using dash::Atomic in DASH containers is not supported, yet.
We will try to work out a solution for your use case. 
I bounced your mail to the developer list (support@dash-project.org) and created an official feature request:

<https://github.com/dash-project/dash/issues/249>

We give feature requests from users high priority. 
Please follow us on GitHub and leave a comment on the issue so wecan keep you up to date.

Let me (or the DASH team on GitHub) know if there is anything we can do to help.

**Comment by rkowalewski** [comment]
I am on it and prepare a small unit test to imitate a scenario as described above.

**Comment by devreal** [comment]
Maybe I'm wrong but to me his use case sounds more like a `SharedCounter`? He wants local updates and eventually combine them into a sum...

**Comment by BenjaProg** [comment]
Hm i thought unordered map ist perfekt, because the mapped value to a key is the counter and beforehand it's unknown which keys are to be found at all.
If I interpret your SharedCounter suggestion right, one need a SharedCounter for every possible Key, but the range in this case would be to large. Or did I miss something?
EDIT: @rkowalewski yes that was perfect timing!

**Comment by rkowalewski** [comment]
hmm..to me it sounds like he wants to count the number of occurences of a particular DNA stripe. So obviously he wants to manage all the stripes (as keys) in an unordered map and increment it if was found found by any unit. So it is some kind of `dash::unodered_map<dash::SharedCounter>`.

EDIT: Actually, @BenjaProg and me submitted the same comment concurrently.

**Comment by fuchsto** [comment]
@devreal @rkowalewski @BenjaProg

Nonono, be careful:

As the name suggests, `dash::SharedCounter` already is a (shared) distributed type, implemented as a wrapper for `dash::Array` so that's not practicable for mapped values in `dash::UnorderedMap`.

Also, with the default hash function, units have to update remote counter values in this use case.
Keys are mapped to a single specific unit, so local matches with a key mapped to a remote unit requires a counter update at this remote unit but `dash::SharedCounter` only supports local updates.

**Comment by devreal** [comment]
@fuchsto You're right, distributed map with distributed arrays as values does not make much sense. If all `GATCSequence`s are known before, a simple `std::unordered_map` with `SharedCounter` would work, though. 

What's the road-block in using `dash::Atomic` in a `dash::unordered_map`?

**Comment by fuchsto** [comment]
We briefly discussed this in the Munich team meeting today and it looks like just some minor changes (changing a member type from `const *` to DART id) are needed to use  `dash::Atomic` as a mapped type in `dash::UnorderedMap`
It is just a wrapper for a `dart_gptr_t` so `dash::Atomic` could even be sold to the compiler as trivially copyable. From the container's perspective, there is nothing special about it.

We will crotchet a proof of concept in the dash-apps repository and see where we get from there.

About using `std::unordered_map`:
If matched sequences are known a priori, the use case is trivial but would not represent real-world pattern matching problems. Even classic "academic" problems like longest common subsequence have no known match at the beginning.

**Comment by fmoessbauer** [comment]
On the compiler side, the changes are minimal. However my solution seems to be not correct anymore, as a `dash::Atomic().add(1)` which is not local leads to DART errors in the accumulate phase, as the DART segment id is wrong. According to all documentation of DART that I found (pretty few^^) this should definitely work. I also checked, that both the DART team ID as well as the dart_globptr is copied correctly (bitwise equal).

Maybe someone familiar with the DART stuff might have a look at the implementation:
https://github.com/dash-project/dash/compare/feat-249-atomic

**Comment by fuchsto** [comment]
(told you so)

Seriously, this is a drastic defect. I will test your use case on some ancient stable code base and check if it is related to recent changes.

**Comment by ddiefenthaler** [comment]
@fmoessbauer To which use case are you referring?
If you're asking about the test ([AtomicTest.cc](https://github.com/dash-project/dash/blob/feat-249-atomic/dash/test/AtomicTest.cc#L115)) it is probably because the default constructed `dash::Atomics` are not valid.
It looks like its writing the values to `DART_GPTR_NULL`, which is done twice. Therefore the comaprision in the test yields 8 instead of 4.

However I don't get any errors when running only the test from the AtomicTest suite.
This is a little bit surprising as I normally would not expect to be allowed to write to `DART_GPTR_NULL`.

**Comment by fmoessbauer** [comment]
@ddiefenthaler exactly, you are right with everything you mentioned. However this should be asserted. Maybe the assertions do not work properly.

After have slept on it for one night I see that simply my unit test was not valid. However this will happen to most of our users, as the behavior of `dash::Atomic` differs massively from `std::Atomic`. While the dash variant is more an algorithm to perform atomic operations on memory, the std variant is a datastructure (with a single member).

We should discuss how to proceed from now: I recommend the following:
- rename `dash::Atomic` to something like `dash::AtomicWrapper`
- introduce a native `dash::Atomic` which behaves like `std::Atomic`

**Comment by devreal** [comment]
I looked into this this morning as well. I am about to add sanity checks in DART communication routines to catch this type of errors. However, the unit IDs and segment IDs passed to DART seem strange, likely uninitialized (not -1 and 0 as would be expected for `DART_GPTR_NULL`). I still cannot track them down.

**Comment by devreal** [comment]
> We should discuss how to proceed from now: I recommend the following:
> 
>     - rename dash::Atomic to something like dash::AtomicWrapper

I'm not a big fan of the name `AtomicWrapper` although I am having a hard time coming up with a good alternative. Maybe `AtomicAccess` or `AtomicAccessor`? 

>     - introduce a native dash::Atomic which behaves like std::Atomic
> 

You mean the new `dash::Atomic` will allocate global memory non-collectively on the current unit and use `dash::AtomicWrapper` internally?

**Comment by devreal** [comment]
One more thought: Do we actually need a dedicated `dash::Atomic` class? As far as I can see, the support for atomic operations does not rely on any internal state but can be executed on any suitable (aka basic) data type through a global pointer. We could thus have a `dash::atomic::` namespace that contains `dash::atomic::fetch_and_op` and `dash::atomic::accumulate`, right? 

Having a wrapper type for atomic operations only makes sense if we want to provide atomic operations on data types other than the basic ones, for which we need an internal state, e.g., to use `dart_lock_*` for mutual exclusion. Do we want to support this kind of extended atomic operations?

**Comment by rkowalewski** [comment]
I agree with @devreal to provide atomic algorithms which operate on global pointers. This simplifies many details and the semantics are well-defined.
For more complex types, I would not provide this out of the box as a wrapper because performance drastically goes down. Instead I would suggest to provide a `dash::mutex(Team & team)` and the user can then aquire and release it to guarantee atomic updates.

**Comment by fuchsto** [comment]
Ah, no, we are leaving the terrain of well-defined concepts/semantics here.
Yes, it would work to use atomic wrappers and atomic operations, but there are strong arguments against doing so.

For starters, I disagree with @rkowalewski on
> This simplifies many details and the semantics are well-defined.

With the suggestions above, they are exactly not.

So for example, you would define a variable `GlobPtr<T> reg`.
One unit could use a register/address for atomic operations and another unit/context could use regular reads/writes on the same variable.
Same variable, same type but: different memory model, different consistency, different operations -> semantics are not well-defined.

Even worse: you could never be sure if a `dart_gptr_t` / `GlobPtr` / ... returned from any operation
requires atomic operations for its use (and no, adding a Doxygen comment still is no solution).

It is crucial to denote atomic variables with specific types. Because types are the only way to couple and validate semantics.

The considerations in C11/C++11 also apply to DASH. It's the same situation: actually you don't need `std::atomic`, why not just use C11 atomic operations on a variable directly?

**You can, but you have to use specific atomic variable types like `atomic_int`.**

<http://en.cppreference.com/w/c/atomic>

Is there any difference between `atomic_int` and `int` in memory / registers or anywhere else apart from the type system?
No. `CMPXCHG` operates on `$al`, `$ax`, `$eax` and doesn't care.
But: when called, the memory model of the registers change (the. memory. model. changes.), escalating to pipeline flushes and eviction of all cache lines associated with these registers (so: *everything* behaves different. Time dilation, physical constants, your phone number) just because a specific instruction has been used.

So as @devreal wrote about `dash::Atomic`:

> the support for atomic operations does not rely on any internal state but can be executed on 
> any suitable (aka basic) data type 

I hope you see that this is true, but irrelevant.

My full remarks on this would be rather wordy and I'm sure there already are good articles in the web on this.

The summary on cppreference.com covers the essentials on atomic semantics and their effects on ... everything. I can provide additional references if needed.

**Comment by ddiefenthaler** [comment]
Thanks for the clarification @fuchsto,
however I do not fully agree with that:
>> the support for atomic operations does not rely on any internal state but can be executed on
any suitable (aka basic) data type
>
> I hope you see that this is true, but irrelevant.

The original use case was to count up some values in a way that no updates are lost. If we are able to achieve this with functions already available in DART there probably should be a way in DASH, too. And that without restricting the access to a bare minimum. I agree that this then should not be called atomic anymore.

Apart from that the current implentation of `dash::Atomic` is flawed in the same way, as the global pointer passed to `dash::Atomic` remains valid. For example in the [AtomicTest](https://github.com/dash-project/dash/blob/f77af766358ce3f9edb2ee1c553349994afdf2a4/dash/test/AtomicTest.cc#L44) suit we can still access the variable `shared`.

So I would suggest to first clarify what exactly is meant by an atomic variable in a distributed environment and how a container filled with atomic variables should look like in DASH.

**Comment by fuchsto** [comment]
No! The DART functions are already malformed in the first place.

There really is no room to discuss alternatives to "make it work somehow", the principles described in my last comment are of rather formal nature and don't result from some implementation-specific mechanism that someone could redesign or abstract away.
Really, there is no excuse.

Yes, the current implementation of `dash::Atomic` is not entirely robust against provoking inconsistent states but that's not a reason to make it worse. It can be misused, but its concept gets the abstraction perfectly right.
The fix for the atomic-pointer-aliasing is trivial, actually: make the `GlobPtr` constructor parameter move-only. `std::atomic` can be easily abused, too, so I actually don't care.

Also, there is not much to clarify, the STL does it perfectly well. There are **additional** restrictions and phenomena in PGAS (think of those as another dimension in consistency and coherency), but the specification from C11/C++11 for atomic operations, memory models, etc. is the minimum requirement for correctness.

I just don't support / satisfy these specifications, yet, because I just needed it for a few use cases.
That's not to say we should ignore fundamental laws of parallel computing.

**Comment by fuchsto** [comment]
@fmoessbauer 

>dash::Atomic differs massively from std::Atomic. While the dash variant is more an algorithm to perform atomic operations on memory, the std variant is a datastructure (with a single member).

No, absolutely not. `std::atomic` is not a container, by no definition. Just look at its signatures.
The pointer member in `std::atomic` is ... a *member*. Not an element.

And neither is `dash::Atomic`. Both are templates for atomic types.
Much like templates for, say, floating point types.

<http://en.cppreference.com/w/cpp/atomic/atomic>

> Each instantiation and full specialization of the std::atomic template defines an atomic type. If one thread writes to an atomic object while another thread reads from it, the behavior is well-defined (see memory model for details on data races)

> std::atomic may be instantiated with any TriviallyCopyable type T. std::atomic is neither copyable nor movable.

This is what `dash::Array` should mimic as closely as possible (I don't care about the non-copyable, non-moveable part for now because MPI has no problem with aliasing) but it requires some additional meta data so I can't define it that "pure".

Here is the full standardese:
<http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2427.html>

**Comment by rkowalewski** [comment]
@fuchsto:
> So for example, you would define a variable GlobPtr<T> reg.
One unit could use a register/address for atomic operations and another unit/context could use regular reads/writes on the same variable.
Same variable, same type but: different memory model, different consistency, different operations -> semantics are not well-defined.

I understand your point, however what I meant with well-defined is that the user defines the semantics. Or to be more precisely the user explicitly requests it. Let's compare it to C11 atomics. You can define any object to be atomic. If we want to load a pointer of an atomic object we have to use `val = atomic_load(my_ptr)`. However, this does not prevent the user to use native loads, such as `val = *my_ptr`. So it is not the memory model's fault but the user's failure.

And if we discuss about the memory model. We do not really have a specified memory model right now. We simply rely on MPI's semantics which are quite limited and rely only on a weak memory model. In contrast C++ guarantees sequential consistency for data-race-free programs. This is something we cannot adopt in our distributed memory architecture.

**Comment by fmoessbauer** [comment]
I agree with @rkowalewski 

The problem with the `dash::Atomic` is that it holds a globptr as member instead of the value itself. However the globptr is necessary to perform the atomic DART operations on the value (which can reside anywhere in the global address space). This indirection is especially problematic, as the value is not stored in the memory of the container. So what about using phantom types like the following:

```
dash::Atomic<T> : T // store value of native type
dash::atomic::add(dash::GlobRef<Atomic<T>> & ref) // only takes atomics
// dash::Atomic<T>().get() => dash::atomic::get(*this);
```

The only problematic point that remains is that it is still possible to construct an `dash::Atomic` outside the global address space. However we could fix this by making the constructor private and use friend declarations for the containers.

@fuchsto IMO the atomic DART operations are totally correct, as they rely on the consistency guarantees of MPI.

**Comment by fuchsto** [comment]
Okay, I see the problem.

`dash::Atomic` does not have to be a non-moveable, non-copyable type. This is required for `std::atomic` to ensure that an atomic variable has
 
- exactly one owner
- that cannot change in the life time of the variable

We have the same requirements for `dash::Atomic`, but they are not realized as C++ type constraints (like copyable, moveable).

Instead of a register, the referenced atomic primitive is `dart_gptr_t` in DASH.
So we have to ensure that:

- the global pointer in `dash::Atomic` references a global address that has exactly one owner
- ... that does not change in the lifetime of the `dash::Atomic`.

We already have this guarantee from DART.
You see my point?

**Comment by fuchsto** [comment]
@fmoessbauer referring to your last comment:

Yes, this is could turn into something we can agree on!

But: We cannot assume MPI to argue soundness of the DART interface because MPI is an arbitrary implementation, so I don't take this as an argument for itself.

**Comment by fmoessbauer** [comment]
Ok, perfect. So what about changing the interface the following way:

- introduce phantom type `dash::Atomic<T>`
- add GlobRef traits to check if type is a GlobRef
- provide algorithms like `dash::atomic::load(ref)` which are only enabled for GlobRefs of atomic types (checked by traits)
- Add specializations of GlobRef for atom types to do something like `array<Atomic<T>>[i].load()`

IMO this will lead to a clean interface.

**Comment by fuchsto** [comment]
Yes! We also need traits for the PGAS memory model, then we should end up with a well-defined
abstraction.

I will create an article in the doc-wiki and describe PGAS atomic semantics for discussion.

**Comment by fuchsto** [comment]
Another remark, back on topic: 

Using `dash::Container<dash::Atomic<T>>` is in fact a practicable approach.
An instance of `dash::Atomic<T>` acts as a **reference** to an atomic variable.
This, again, is why cloning instances of `dash::Atomic` is okay.

This would correspond to `std::container<std::atomic<T> &>` which is not allowed for STL containers (`value_type` must not be a reference), but perfectly fine for DASH containers.

This is analog to `dash::Container<dash::GlobRef<T>>` which would not raise any concern.
Or: does it?

**Comment by fuchsto** [comment]
@rkowalewski 

>  We simply rely on MPI's semantics which are quite limited and rely only on a weak memory model.

Right!

> In contrast C++ guarantees sequential consistency for data-race-free programs. This is something we cannot adopt in our distributed memory architecture.

Yes, but where is the problem with that? So we don't provide `memory_order_seq_cst`, right, but we still can specify any other possible memory order.
We do not and cannot satisfy STL semantics **entirely**, but we don't have to.

**Comment by devreal** [comment]
While implementing `dart_compare_and_swap` and subsequently `compare_exchange`, I realized that the DART atomic operations take a `teamid` parameter, which is neither used nor required. This makes it necessary that `dash::AtomicReference` stores the  team ID. Was this a deliberate decision, maybe required for the shmem communication backend? If not, I would simply remove the parameter as it wrongly suggests to the user that the atomic operations are collective.

**Comment by fuerlinger** [comment]
I don't remember this being a deliberate decision. I think it was just an oversight so I think we can remove the `teamid` parameter.

**Comment by fuchsto** [comment]
The idea behind the team parameter was to reflect the dependency on team-allocated memory.
That is to say: the global address is only atomic within a specific team.
But this is an unsurprising restriction that doesn't need to be explicitly stated.
I'm with @devreal, it implies a collective operation so we should remove it.

**Comment by rkowalewski** [comment]
As you can see in the linked issue I will implement the use case from scratch in DASH. It should be a very good benchmark for DASH dynamic data structures and memory management.

**Comment by fuchsto** [comment]
Addressed in #255

**Comment by fuchsto** [comment]
The requested use case works with changes in PR #295 but copy/assignment semantics of `dash::Atomic` are still not well-defined, see separate issue #289

**Comment by fuchsto** [comment]
Fixed in merge of #295
